### PR TITLE
Initial feature commit. Added blur.

### PR DESCRIPTION
**Pull request by JamesF2440**
Created a new branch for features to be pushed to. This has all the processor utility, video class has been cleaned up a little bit. Should be good to merge to main to allow for all features to be used.

### CODE REVIEW

File: aEye/processor.py [hunk 1]
```diff
@@ -172,6 +218,180 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        None, but creates cropped video in output folder
+        """
+        for video in self.video_list:
+            cmd = (
+                f"{ffmpeg} -y -i '{video.get_file()}' -v quiet -filter:v 'crop={section_width}:{section_height}:{start_x}:{start_y}' "
+                f"-c:a copy modified/section_crop_{video.title}.mp4"
+            )
+            subprocess.call(cmd, shell=True)
 
+    def moving_crop(self):
```
Comment (vinayakSharm) [inline aEye/processor.py:309]: remove this function from pr. save for future pr, when you can figure out how to add params.

**Comment by vinayakSharm** [comment]
update requirements to include static_ffmpeg

File: aEye/processor.py [hunk 2]
```diff
@@ -172,6 +218,157 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
         print("successfully upload the output files S3 bucket: s3://aeye-data-bucket/modified/")
         print("successfully remove the output file from local machine")
 
+    def trim_video_from_to(self, start, end):
```
Comment (vinayakSharm) [inline aEye/processor.py:221]: rename to "trim_video_start_end"

Comment (vinayakSharm) [reply aEye/processor.py:221]: or trim_video_starttime_endtime etc.

File: aEye/processor.py [hunk 3]
```diff
@@ -172,6 +218,157 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        *Error checking and minute conversion would probably be good...
 
+        Returns
+        -------
+        None, but creates trimmed video in output folder
+        """
+        for video in self.video_list:
+            duration = end - start
+            cmd = f"{ffmpeg} -y -ss {start} -i '{video.get_file()}' -v quiet -t {duration} -c copy " \
+                  f"modified/split_on_time_{video.title}"
```
Comment (vinayakSharm) [inline aEye/processor.py:234]: can you have the output of the video name be the same as the function that was called pls

File: aEye/processor.py [hunk 4]
```diff
@@ -172,6 +218,157 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+
+        Returns
+        -------
+        None, but frame grabbed is displayed in output folder
+        """
+        for video in self.video_list:
+            cv_video = video.cv_video
+            cv_video.set(cv2.CAP_PROP_POS_FRAMES, frame)
+            ret, output = cv_video.read()
+            cv2.imwrite(f"modified/cv2specific_frame_{video.title}.png", output)
```
Comment (vinayakSharm) [inline aEye/processor.py:342]: name the video output the same as the function call pls. ie "extract_specific_frame_{frame}_{video_title}.png

**Comment by vinayakSharm** [comment]
locally saved videos need to be written according to the function call made to them. 1:1 pls

ex/
name the video output the same as the function call pls. ie "extract_specific_frame_{frame}_{video_title}.png

File: README.md [hunk 5]
```diff
@@ -46,9 +46,27 @@ process = Processor()
```
Comment (vinayakSharm) [inline README.md:46]: make it very clear in the readme that the following instructions are for jupyter notebook users. 

later we will be using .py files so we can have another runner section or replace this one. thx

File: aEye/processor.py [hunk 6]
```diff
@@ -157,13 +204,12 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
         """
 
-
         for video in self.video_list:
-
             path = 'modified/output_' + video.title
-            response = self._s3.upload_file( path, bucket,  path)
+            response = self._s3.upload_file(path, bucket, path)
+
+            # This will delete all file from RAM and local machine.
```
Comment (vinayakSharm) [inline aEye/processor.py:211]: @JamesF2440 your current naming convention and this upload method are not compatable. either fix the upload function to upload your videos or rename your files.

Comment (vinayakSharm) [reply aEye/processor.py:211]: also this upload function removes all videos from memory, so there is no way to clear out this folder at the time being.

File: aEye/processor.py [hunk 7]
```diff
@@ -172,6 +218,157 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+            duration = end - start
+            cmd = f"{ffmpeg} -y -ss {start} -i '{video.get_file()}' -v quiet -t {duration} -c copy " \
+                  f"modified/split_on_time_{video.title}"
+            subprocess.call(cmd, shell=True)
 
+    def trim_into_clips(self, interval):
+        """
+        trim_into_clips(interval, file)
+        This method splits a video into *interval* second long clips, but for any remainder
+        the last clip will be truncated. For example: A 43 second long video with a 10 second interval
```
Comment (vinayakSharm) [inline aEye/processor.py:241]: good note. glad you are verbose on the functionality.

File: aEye/processor.py (hunk 2 above)
Comment (vinayakSharm) [inline aEye/processor.py:221]: no logging?

File: aEye/processor.py [hunk 8]
```diff
@@ -172,6 +218,157 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        -------
+        None, but creates trimmed video in output folder
+        """
+        for video in self.video_list:
+            duration = end - start
+            cmd = f"{ffmpeg} -y -ss {start} -i '{video.get_file()}' -v quiet -t {duration} -c copy " \
+                  f"modified/split_on_time_{video.title}"
+            subprocess.call(cmd, shell=True)
 
+    def trim_into_clips(self, interval):
```
Comment (vinayakSharm) [inline aEye/processor.py:237]: no logging?

File: aEye/processor.py [hunk 9]
```diff
@@ -172,6 +218,157 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        None, but creates interval second long videos in output folder
+        """
+        for video in self.video_list:
+            cmd = (
+                f"{ffmpeg} -y -i '{video.get_file()}' -c copy -map 0 -segment_time {interval} "
+                f"-f segment -reset_timestamps 1 modified/interval_split{video.title}%02d.mp4"
+            )
+            subprocess.call(cmd, shell=True)
+
+    def split_on_frame(self, frame):
```
Comment (vinayakSharm) [inline aEye/processor.py:256]: no logging?

**Comment by vinayakSharm** [comment]
pls add logging to all of your functions

File: aEye/processor.py [hunk 10]
```diff
@@ -172,6 +218,157 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        """
+        split_on_frame(frame, video)
```
Comment (vinayakSharm) [inline aEye/processor.py:258]: what does line 258 mean? what does a reader or user gain from reading 258?

File: README.md [hunk 11]
```diff
@@ -46,9 +46,27 @@ process = Processor()
 
 ```console
 process.load_and_resize(bucket = 'aeye-data-bucket', prefix = 'input_video/', x_ratio = .6, y_ratio = .5)
+# To load without resizing:
+process.load(bucket = 'aeye-data-bucket', prefix = 'input_video/')
 ```
 
-6. Upload the result to the desire bucket
+6. Use the Processor to perform operations on video objects
```
Comment (vinayakSharm) [inline README.md:53]: pls include all functions in this section. Including the ones that were built by luan in the previous feature push

File: aEye/video.py [hunk 12]
```diff
@@ -3,78 +3,99 @@
...
+        Returns video duration in seconds, but does so as a string
+    get_frames -> str:
+        Returns the amount of frames in the video as a string (via OpenCV)
+    getfile -> str:
+        Returns video file path
 
     """
 
-    def __init__(self,file , title = None ) -> None:
+    def __init__(self, file, title = None) -> None:
```
Comment (vinayakSharm) [inline aEye/video.py:44]: hi james your video class no longer has an attribute named width. 

this breaks the compatibility of the already existing functions such as resize_by_ratio(x_ratio=.8, y_ratio=.8) which rely on width

File: aEye/processor.py [hunk 13]
```diff
@@ -63,29 +108,29 @@ def load(self, bucket=  'aeye-data-bucket', prefix='input_video/') -> list:
...
             self.video_list.append(Video(url, title))
 
         logging.info(f"Successfully loaded video data from {bucket}")
         logging.info(f"There are total of {len(self.video_list)} video files")
 
         return self.video_list
 
-
-    def resize_by_ratio(self, x_ratio = .8, y_ratio = .8) -> None:
+    def resize_by_ratio(self, x_ratio=.8, y_ratio=.8) -> None:
```
Comment (vinayakSharm) [inline aEye/processor.py:133]: this function broke with the removal of attributes from the video class

File: aEye/processor.py [hunk 14]
```diff
@@ -97,35 +142,37 @@ def resize_by_ratio(self, x_ratio = .8, y_ratio = .8) -> None:
...
+
         """
 
-        #This will loop to the list of videos to apply the resizing feature. 
+        # This will loop to the list of videos to apply the resizing feature.
         for video in self.video_list:
 
-            new_width = int(video.width * x_ratio )
-            new_height = int(video.height * y_ratio )
+            new_width = int(video.width * x_ratio)
```
Comment (vinayakSharm) [inline aEye/processor.py:151]: need video.width

File: aEye/processor.py [hunk 15]
```diff
@@ -172,6 +218,157 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
         print("successfully upload the output files S3 bucket: s3://aeye-data-bucket/modified/")
         print("successfully remove the output file from local machine")
 
+    def trim_video_from_to(self, start, end):
+        """
+        trim_video_from_to(start, end, file)
```
Comment (vinayakSharm) [inline aEye/processor.py:223]: redundant line 233. not needed

File: aEye/processor.py (hunk 13 above)
Comment (vinayakSharm) [reply aEye/processor.py:133]: video class still breaks this function. 

need AttributeError: 'Video' object has no attribute 'cap'

File: README.md
Comment (vinayakSharm) [inline README.md:1]: needs updates

File: aEye/processor.py [hunk 16]
```diff
@@ -172,6 +218,160 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
         print("successfully upload the output files S3 bucket: s3://aeye-data-bucket/modified/")
         print("successfully remove the output file from local machine")
 
+    def trim_video_start_end(self, start, end):
```
Comment (vinayakSharm) [inline aEye/processor.py:221]: I am getting inconsistent naming conventions here. 

some vidoes including the seconds duration in title, others are missing the 's'.

Screen Shot 2023-06-23 at 3 44 01 PM

File: aEye/processor.py [hunk 17]
```diff
@@ -172,6 +218,160 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        None, but creates trimmed video in output folder
+        """
+        for video in self.video_list:
+            duration = end - start
+            cmd = f"{ffmpeg} -y -ss {start} -i '{video.get_file()}' -v quiet -t {duration} -c copy " \
+                  f"modified/output_trim_video_start_end_{video.title}"
+            subprocess.call(cmd, shell=True)
+            logging.info(f"Created a sub-video from {start} to {end}")
+
+    def trim_into_clips(self, interval):
```
Comment (vinayakSharm) [inline aEye/processor.py:237]: wrong naming convention. pls fix 
Screen Shot 2023-06-23 at 3 46 11 PM

File: aEye/processor.py [hunk 18]
```diff
@@ -172,6 +218,160 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        """
+        for video in self.video_list:
+            cmd = (
+                f"{ffmpeg} -y -i '{video.get_file()}' -c copy -map 0 -segment_time {interval} "
+                f"-f segment -reset_timestamps 1 modified/output_trim_into_clips_{video.title}%02d.mp4"
+            )
+            subprocess.call(cmd, shell=True)
+            logging.info(f"Video has been trimmed into {interval} second long clips!")
+
+    def split_on_frame(self, frame):
```
Comment (vinayakSharm) [inline aEye/processor.py:256]: naming wrong. 
Screen Shot 2023-06-23 at 3 53 35 PM

File: aEye/processor.py [hunk 19]
```diff
@@ -172,6 +218,160 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        None, but creates trimmed video in output folder
+        """
+        for video in self.video_list:
+            fps = float(video.get_num_frames()) / float(video.get_duration())
+            time_stamp = start_frame / fps
+            cmd = (
+                f"{ffmpeg} -y -ss {str(time_stamp)} -i '{video.get_file()}' -v quiet -c:v libx264 -frames:v {num_frames}"
+                f" modified/output_split_num_frames_{start_frame}_to_{(start_frame + num_frames)}_{video.title}.mp4"
+            )
+            logging.info("Encoding ", num_frames, " from ", start_frame)
```
Comment (vinayakSharm) [inline aEye/processor.py:292]: logging broken

File: aEye/processor.py [hunk 20]
```diff
@@ -172,6 +218,160 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+            subprocess.call(cmd, shell=True)
+
+    def crop_video_section(self, start_x, start_y, section_width, section_height):
```
Comment (vinayakSharm) [inline aEye/processor.py:295]: naming wrong

File: aEye/processor.py [hunk 21]
```diff
@@ -172,6 +218,160 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        step for ANY filter application, so there will be noticeable processing time.
 
+        Returns
+        -------
+        None, but creates cropped video in output folder
+        """
+        for video in self.video_list:
+            cmd = (
+                f"{ffmpeg} -y -i '{video.get_file()}' -v quiet -filter:v 'crop={section_width}:{section_height}:{start_x}:{start_y}' "
+                f"-c:a copy modified/output_crop_video_section_{video.title}.mp4"
```
Comment (vinayakSharm) [inline aEye/processor.py:308]: adding mp4 here creates duplicate mp4

File: aEye/processor.py [hunk 22]
```diff
@@ -172,6 +218,160 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+            )
+            subprocess.call(cmd, shell=True)
+            logging.info(f"Created a {section_width}x{section_height} crop at ({start_x},{start_y})")
+
+    def cv_extract_frame_at_time(self, time):
```
Comment (vinayakSharm) [inline aEye/processor.py:313]: remove mp4 from naming

File: aEye/processor.py [hunk 23]
```diff
@@ -172,6 +218,160 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        for video in self.video_list:
+            cv_video = video.cv_video
+            fps = cv_video.get(cv2.CAP_PROP_FPS)
+            frame_id = int(fps * time)
+            cv_video.set(cv2.CAP_PROP_POS_FRAMES, frame_id)
+            ret, frame = cv_video.read()
+            cv2.imwrite(f"modified/output_cv_extract_frame_at_time_{time}_{video.title}.png", frame)
+            logging.info(f"Extracted frame at time {time}")
+
+    def cv_extract_specific_frame(self, frame):
```
Comment (vinayakSharm) [inline aEye/processor.py:330]: remove mp4

File: aEye/processor.py [hunk 24]
```diff
@@ -172,6 +218,160 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        None, but frame grabbed is displayed in output folder
+        """
+        for video in self.video_list:
+            cv_video = video.cv_video
+            cv_video.set(cv2.CAP_PROP_POS_FRAMES, frame)
+            ret, output = cv_video.read()
+            cv2.imwrite(f"modified/output_cv_extract_specific_frame_{frame}_{video.title}.png", output)
+            logging.info(f"Frame #{frame} extracted ")
+
+    def extract_many_frames(self, start_frame, num_frames):
```
Comment (vinayakSharm) [inline aEye/processor.py:345]: naming wrong --> mp404.mp4 wierd

File: aEye/processor.py [hunk 25]
```diff
@@ -172,6 +218,160 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        -------
+        None, but frames are displayed in output folder
+        """
+        for video in self.video_list:
+            cmd = f"{ffmpeg} -i '{video.get_file()}' -vf select='eq(n\,{start_frame})' -frames:v {num_frames} " \
+                  f"modified/output_extract_many_frames_{video.title}%02d.png"
+            subprocess.call(cmd, shell=True)
+            logging.info(f"Extracted {num_frames} from video, saved as PNG's")
+
+    def blur_video(self, blur_level, blur_steps=1):
```
Comment (vinayakSharm) [inline aEye/processor.py:361]: naming wrong. pls fix

Comment (vinayakSharm) [inline aEye/processor.py:361]: docstrings need inputs described

File: aEye/processor.py (hunk 24 above)
Comment (vinayakSharm) [inline aEye/processor.py:345]: docstrings need inputs described

File: aEye/processor.py (hunk 23 above)
Comment (vinayakSharm) [inline aEye/processor.py:330]: docstrings need inputs described

File: aEye/processor.py (hunk 22 above)
Comment (vinayakSharm) [inline aEye/processor.py:313]: docstrings need inputs described

File: aEye/processor.py (hunk 20 above)
Comment (vinayakSharm) [inline aEye/processor.py:295]: docstrings need inputs described

File: aEye/processor.py [hunk 26]
```diff
@@ -172,6 +218,160 @@ def upload(self, bucket=  'aeye-data-bucket') -> None:
...
+        """
+        for video in self.video_list:
+            fps = float(video.get_num_frames()) / float(video.get_duration())
+            time_stamp = frame / fps
+            cmd = f"{ffmpeg} -y -ss {time_stamp} -i '{video.get_file()}' -v quiet -c:v libx264 -c:a aac" \
+                  f" modified/output_split_on_frame_{video.title}.mp4"
+            subprocess.call(cmd, shell=True)
+            logging.info(f"Split video at frame {frame}")
+
+    def split_num_frames(self, start_frame, num_frames):
```
Comment (vinayakSharm) [inline aEye/processor.py:274]: docstrings need inputs described

File: aEye/processor.py (hunk 18 above)
Comment (vinayakSharm) [inline aEye/processor.py:256]: docstrings need inputs described

File: aEye/processor.py (hunk 17 above)
Comment (vinayakSharm) [inline aEye/processor.py:237]: docstrings need inputs described

File: aEye/processor.py (hunk 16 above)
Comment (vinayakSharm) [inline aEye/processor.py:221]: docstrings need inputs described

File: aEye/video.py [hunk 27]
```diff
@@ -3,78 +3,114 @@
...
-    def cleanup(self) -> None:
+    def get_file(self):
         """
-        This method will release the current view of video object from RAM.
+        Returns the file path as a string
         """
-        self.capture.release()
+        return self.file
+
+    def get_width(self):
```
Comment (vinayakSharm) [inline aEye/video.py:103]: docstring

File: aEye/video.py [hunk 28]
```diff
@@ -3,78 +3,114 @@
...
+        if self.meta_data is not None:
+            return self.meta_data["streams"][0]["width"]
+        else:
+            self.meta_data = self.extract_metadata()
+            return self.meta_data["streams"][0]["width"]
+
+    def get_height(self):
```
Comment (vinayakSharm) [inline aEye/video.py:110]: docstring

**Review by vinayakSharm — DISMISSED** [review]
existing processing function seems to be broken. 

all functions have incorrect naming conventions. pls name remove any wierd extensions. 
pls include params used to create the video as part of the name. 

docstrings need work

**Comment by vinayakSharm** [comment]
need updated readme

File: aEye/video.py [hunk 29]
```diff
@@ -3,68 +3,41 @@
...
-
-
-        key: string
-            The key that associates with the video file.
-
-
-
-
-
+    file: str
```
Comment (PierceLovesee) [inline aEye/video.py:21]: Please add a blank line between each attribute and method.  This helps a lot with readability.  

for example:

## Do this:

```python
"""

method 1(input1, input2) -> output_datatype[output_name]:
    description of method1

method 2(input1, input2) -> output_datatype[output_name]:
    description of method2
"""
```

##  Not this:

```python
"""

method 1(input1, input2) -> output_datatype[output_name]:
    description of method1
method 2(input1, input2) -> output_datatype[output_name]:
    description of method2
"""
```

File: aEye/video.py [hunk 30]
```diff
@@ -3,68 +3,41 @@
...
-
-        add_label(self, mod) -> None:
-            Add ffmpeg label to video object.
-
-        reset_label() -> None:
-            Reset and remove all labels.
-
-        get_label(self) -> string:
-            Get ffmpeg label from video objects.
+    ----------
```
Comment (PierceLovesee) [inline aEye/video.py:29]: add some blank lines between methods for readability.

File: aEye/video.py [hunk 31]
```diff
@@ -160,19 +205,56 @@ def add_label(self, label):
         """
         self.label += label
 
+    def add_output_title(self, title):
+        """
```
Comment (PierceLovesee) [inline aEye/video.py:209]: docstring.

File: aEye/video.py [hunk 32]
```diff
@@ -160,19 +205,56 @@ def add_label(self, label):
...
+        self.label = ''
 
     def get_label(self):
         """
         This method will return the all ffmpeg label from the video.
         """
         return self.label
 
+    def set_output(self, new_out):
+        """
```
Comment (PierceLovesee) [inline aEye/video.py:227]: docstring

File: aEye/video.py [hunk 33]
```diff
@@ -3,69 +3,46 @@
...
-
-
-        key: string
-            The key that associates with the video file.
-
-
-
-
-
+    file: str
```
Comment (PierceLovesee) [inline aEye/video.py:26]: Was the correct version of this docstring kept?

File: aEye/video.py [hunk 34]
```diff
@@ -161,19 +214,54 @@ def add_label(self, label):
         """
         self.label += label
 
+    def add_output_title(self, title):
+        """
```
Comment (PierceLovesee) [inline aEye/video.py:218]: Was the correct version of this docstring kept?

File: aEye/video.py [hunk 35]
```diff
@@ -161,19 +214,54 @@ def add_label(self, label):
...
+        self.label = ''
 
     def get_label(self):
         """
         This method will return the all ffmpeg label from the video.
         """
         return self.label
 
+    def set_output(self, new_out):
+        """
```
Comment (PierceLovesee) [inline aEye/video.py:236]: Was this correct version of this docstring kept?

**Review by PierceLovesee — APPROVED** [review]

**Pull request closed without merging**